esptool.py

```diff
@@ -2433,15 +2470,28 @@ def add_spi_flash_subparsers(parent, is_elf2image):
 
     if operation_args[0] == 'esp':  # operation function takes an ESPLoader connection object
         initial_baud = min(ESPLoader.ESP_ROM_BAUD, args.baud)  # don't sync faster than the default baud rate
-        if args.chip == 'auto':
-            esp = ESPLoader.detect_chip(args.port, initial_baud, args.before, args.trace)
-        else:
-            chip_class = {
-                'esp8266': ESP8266ROM,
-                'esp32': ESP32ROM,
-            }[args.chip]
-            esp = chip_class(args.port, initial_baud, args.trace)
-            esp.connect(args.before)
+        ser_list = all_serial_ports_2() if args.port is None else [args.port]
+        ser_attempts = 0
+        for each_port in reversed(ser_list):
+            ser_attempts += 1
+            if args.port is None:
+                print("Trying %s" % each_port)
+            try:
+                if args.chip == 'auto':
+                    esp = ESPLoader.detect_chip(each_port, initial_baud, args.before, args.trace)
+                else:
+                    chip_class = {
+                        'esp8266': ESP8266ROM,
+                        'esp32': ESP32ROM,
+                    }[args.chip]
+                    esp = chip_class(each_port, initial_baud, args.trace)
+                    esp.connect(args.before)
+                break
+            except FatalError as err:
+                if ser_attempts >= len(ser_list) or args.port is not None:
```

Comment:
This is a bit of an awkward way to check if you're out of options, and also it swallows the exception object details (stack trace, etc) for a failed connection (which may make debugging harder for a specified port). How about something like:

```
except FatalError as err:
   if args.port is not None:
       raise
   print("%s failed to connect: %s" % (each_port, err))
   esp = None

then, outside the loop...

```
if esp is None:
   raise FatalError("None of %d available serial ports could connect." % len(ser_list))
```

Reply:
done